File: 3743-reschedule-meetings-for-maximum-free-time-i/reschedule-meetings-for-maximum-free-time-i.cpp
class Solution {
public:
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        int n = startTime.size();
        int start = 0, end = 0;
        int prev_end = 0;

        int max_free_time = 0, free_time = startTime[0];
        while (end < n) {
            while (end < n && end - start < k) {
                end++;
                if (end < n)
                    free_time += startTime[end] - endTime[end - 1];
            }
            if (end == n)
                free_time += eventTime - endTime[end - 1];
            max_free_time = max(max_free_time, free_time);
            free_time -= startTime[start] - prev_end;
            prev_end = endTime[start];
            start++;
        }

        return max_free_time;
    }
};

File: 3743-reschedule-meetings-for-maximum-free-time-i/reschedule-meetings-for-maximum-free-time-i.java
class Solution {
    public int maxFreeTime(int eventTime, int k, int[] startTime, int[] endTime) {
        int n = startTime.length;

        int[] gaps = new int[n + 1];
        gaps[0] = startTime[0];
        gaps[n] = eventTime - endTime[n - 1];
        for (int i = 0; i < n - 1; i++) {
            gaps[i + 1] = startTime[i + 1] - endTime[i];
        }

        int window = 0, result = 0;
        for (int i = 0; i < k; i++) {
            window += gaps[i];
        }
        for (int i = k; i <= n; i++) {
            window += gaps[i];
            result = Math.max(result, window);
            window -= gaps[i - k];
        }

        return result;



    }
}